tachyon/zk/plonk/vanishing/prover_vanishing_argument.h

```diff
@@ -99,18 +90,31 @@ VanishingConstructed<Poly> CommitFinalHPoly(
       base::CreateVector(quotient_poly_degree,
                          [prover]() { return prover->blinder().Generate(); });
 
-  return {std::move(h_pieces), std::move(h_blinds), std::move(committed)};
+  return {std::move(h_poly), std::move(h_blinds), std::move(committed)};
 }
 
-template <typename PCS, typename Poly, typename F, typename Commitment>
+template <typename PCS, typename Poly, typename ExtendedPoly, typename F,
+          typename Commitment>
 VanishingEvaluated<Poly> CommitRandomEval(
-    const PCS& pcs, VanishingConstructed<Poly>&& constructed, const F& x,
-    const F& x_n, crypto::TranscriptWriter<Commitment>* writer) {
+    const PCS& pcs, VanishingConstructed<Poly, ExtendedPoly>&& constructed,
+    const F& x, const F& x_n, crypto::TranscriptWriter<Commitment>* writer) {
   using Coeffs = typename Poly::Coefficients;
 
-  Poly h_poly = Poly::template LinearCombination</*forward=*/false>(
-      constructed.h_pieces(), x_n);
+  size_t n = pcs.N();
+  auto h_chunks =
+      base::Chunked(constructed.h_poly().coefficients().coefficients(), n);
+  std::vector<absl::Span<F>> h_pieces =
+      base::Map(h_chunks.begin(), h_chunks.end(),
+                [](absl::Span<F> h_piece) { return h_piece; });
+  std::vector<F> coeffs(n);
+  for (size_t i = h_pieces.size() - 1; i != SIZE_MAX; --i) {
+    OPENMP_PARALLEL_FOR(size_t j = 0; j < n; ++j) {
+      coeffs[j] *= x_n;
+      coeffs[j] += h_pieces[i][j];
+    }
+  }
```

Comment:
Can't we use `OPENMP_PARALEL_NESTED_FOR` here?

Reply:
We can't use it here because the computation of the current rows depends on the computation of the previous rows.